unicode_regex.h

```diff
@@ -0,0 +1,252 @@
+#pragma once
+
+#include "unicode.h"
+#include "unordered_set"
+
+class llm_regex {
+public:
+    std::vector<std::string> gpt2_style(const std::string & str) {

```

Comment:
These implementations should be put behind a common API that accepts a regex string. Something like:

```c++
std::vector<std::vector<uint32_t>> regex_split(const std::string & str, const std::string & regex);
```

We then check if the regex is one that is implemented - if not we throw an error.
We should add tests where we compare the outputs from `regex_split` with reference Python implementations